foursight_core/react/api/auth.py

```diff
@@ -0,0 +1,203 @@
+import boto3
+import time
+from .cookie_utils import read_cookie
+from .envs import Envs
+from .jwt_utils import jwt_decode, jwt_encode
+
+
+class Auth:
+
+    def __init__(self, auth0_client_id: str, auth0_secret: str, envs: Envs):
+        self.auth0_client_id = auth0_client_id
+        self.auth0_secret = auth0_secret
+        self.envs = envs
+
+    cache_aws_credentials = {}
+
+    def authorize(self, request: dict, env: str) -> dict:
+        """
+        Verifies that the given request is authenticated AND authorized, based on the authtoken
+        cookie (a JWT-signed-encoded value) in the given request. If so, returns the verified and
+        decoded authtoken as a dictionary. If not, returns a dictionary indicating not authorized
+        and/or not authenticated, and the basic info contained in the authtoken.
+        """
+        try:
+
+            # Read the authtoken cookie.
+
+            authtoken = read_cookie(request, "authtoken")
+            if not authtoken:
+                return self._create_not_authenticated_response(request, "no-authtoken")
+
+            # Decode the authtoken cookie.
+
+            authtoken_decoded = self.decode_authtoken(authtoken)
+            if not authtoken_decoded:
+                return self._create_not_authenticated_response(request, "invalid-authtoken")
+
+            # Sanity check the decoded authtoken.
+
+            if authtoken_decoded["authorized"] is not True or authtoken_decoded["authenticated"] is not True:
```

Comment:
Do we really want to not distinguis unauthorized from unauthenticated? These seem vastly different. Is this blurring part of some externally-dictated protocol?

Also, unless there is some third value besides `True` or `False` that `'authorized'` or `'authenticated'` take on, then `x is True` is the same as `x`, and `x is not True` is the same as `not x`. You only need the longer construction if there's some third value that has to be distinguished. If there _is_ such a third value, you should add a comment indicating that this is what's going on.

```suggestion
            if not authtoken_decoded["authorized"] or not authtoken_decoded["authenticated"]:
```

Reply:
Yes, unauthorized from unauthenticated are very different, and we do distinguish (HTTP 403 for the former, and 401 for the latter). This is where that confusing function comes into play that you commented on below (_create_not_authorized_response) ... Hmm, actually, this may be a corner case, I need to look at this bit more ... I'll get back to you.

WRT the testing for True though, the reason I'm doing that is just to make sure it really is in fact a boolean type *and* True, I don't want it to evaluate to True if it is some other non-boolean but truthy value ... Is there a better way to do that?

Reply:
No, that's the way to say that, but it's unusual and I'd like to understand why such a value would not be true.

Reply:
Okay, getting back to this, I realized that we actually do not need the "authorized" property in the authtoken cookie. If someone is not authorized (not allowed to access a specified environment) they may still be authenticated (logged in); the check of whether or not they are authorized is done on a per-page/API-request basis and is based on the specified environment (and of course first on whether or not they are authenticated), but the check of whether or not they are authenticated (also done on a per-page/API-request basis) is based on the authtoken cookie. So the authtoken does not need an authorized property.

Reply:
And WRT the checking explicitly for is True, the reason is because the data we're looking at is from a cookie in the request, and conceivably, a hacker could possibly (possible in the sense that it wouldn't violate the laws of physics - _highly_ unlikely as the the hacker would need our secret token to encrypt (JWT-sign/encode) the cookie) change the cookie in any number of ways, and so I thought explicitly checking the types and the values both was prudent.